usr/lib/pkcs11/ica_s390_stdll/ica_specific.c

```diff
@@ -943,1840 +942,1837 @@ CK_RV token_specific_sha_final(STDLL_TokData_t *tokdata, DIGEST_CONTEXT *ctx,
+                                   CK_ULONG *out_data_len, OBJECT *key_obj)
+{
+    CK_RV rc;
+    CK_BYTE data[MAX_RSA_KEYLEN], sig[MAX_RSA_KEYLEN];
+    CK_BBOOL flag;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_ULONG modulus_bytes;
+
+    flag = template_attribute_find(key_obj->template, CKA_MODULUS, &attr);
+    if (flag == FALSE) {
+        TRACE_ERROR("Could not find CKA_MODULUS for the key.\n");
+        return CKR_FUNCTION_FAILED;
+    } else {
+        modulus_bytes = attr->ulValueLen;
+    }
+
+    // prepad with zeros
+    //
+    memset(data, 0x0, modulus_bytes - in_data_len);
+    memcpy(&data[modulus_bytes - in_data_len], in_data, in_data_len);
+
+    rc = os_specific_rsa_decrypt(data, modulus_bytes, sig, key_obj);
+    if (rc == CKR_OK) {
+        memcpy(out_data, sig, modulus_bytes);
+        *out_data_len = modulus_bytes;
+    } else {
+        TRACE_DEVEL("os_specific_rsa_decrypt failed\n");
+    }
+
+    return rc;
+}
+
+
+CK_RV token_specific_rsa_x509_verify(STDLL_TokData_t *tokdata,
+                                     CK_BYTE *in_data, CK_ULONG in_data_len,
+                                     CK_BYTE *signature, CK_ULONG sig_len,
+                                     OBJECT *key_obj)
 {
-	CK_RV		rc;
-	CK_BYTE		clear[MAX_RSA_KEYLEN], cipher[MAX_RSA_KEYLEN];
-	CK_ULONG	modulus_bytes;
-	CK_BBOOL	flag;
-	CK_ATTRIBUTE	*attr = NULL;
-
-	flag = template_attribute_find(key_obj->template, CKA_MODULUS, &attr);
-	if (flag == FALSE) {
-		TRACE_ERROR("Could not find CKA_MODULUS for the key.\n");
-		return CKR_FUNCTION_FAILED;
-	} else
-		modulus_bytes = attr->ulValueLen;
-
-	/* format the data */
-	rc = rsa_format_block(tokdata, in_data, in_data_len, clear,
-			      modulus_bytes, PKCS_BT_2);
-	if (rc != CKR_OK) {
-		TRACE_DEVEL("rsa_format_block failed\n");
-		return rc;
-	}
-
-	rc = os_specific_rsa_encrypt(clear, modulus_bytes, cipher, key_obj);
-	if (rc == CKR_OK) {
-		memcpy(out_data, cipher, modulus_bytes);
-		*out_data_len = modulus_bytes;
-	} else
-		TRACE_DEVEL("os_specific_rsa_encrypt failed\n");
-
-	return rc;
+    CK_RV rc;
+    CK_BYTE out[MAX_RSA_KEYLEN];
+    CK_BBOOL flag;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_ULONG modulus_bytes;
+
+    flag = template_attribute_find(key_obj->template, CKA_MODULUS, &attr);
+    if (flag == FALSE) {
+        TRACE_ERROR("Could not find CKA_MODULUS for the key.\n");
+        return CKR_FUNCTION_FAILED;
+    } else {
+        modulus_bytes = attr->ulValueLen;
+    }
+
+    rc = os_specific_rsa_encrypt(signature, modulus_bytes, out, key_obj);
+    if (rc == CKR_OK) {
+        CK_ULONG pos1, pos2, len;
+
+        // it should be noted that in_data_len is not necessarily
+        // the same as the modulus length
+        //
+        for (pos1 = 0; pos1 < in_data_len; pos1++)
+            if (in_data[pos1] != 0)
+                break;
+
+        for (pos2 = 0; pos2 < modulus_bytes; pos2++)
+            if (out[pos2] != 0)
```

Comment:
There is a break missing after the if. You seem to have missed the fix from commit 02a5840afef2fd3b5879d4bed35e17c809341cf0 (ICA: Fix token_specific_rsa_x509_verify() function)

Reply:
Also the indention of the comment and the code block following is wrong. It needs to be 8 positions to the left (aligned with "len = in_data_len - pos1;" right after it)